test/functional/feature_pop_activate.py

```diff
@@ -0,0 +1,107 @@
+        self.num_nodes = 2
+        self.extra_args = [["-txindex"], ["-txindex"]]
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+        self.skip_if_no_pypopminer()
+
+    def setup_network(self):
+        self.setup_nodes()
+
+    def get_best_block(self, node):
+        hash = node.getbestblockhash()
+        return node.getblock(hash)
+
+    def _cannot_endorse(self):
+        self.log.warning("starting _cannot_endorse()")
+
+        # node0 start with 100 blocks
+        self.nodes[0].generate(nblocks=100)
+        self.log.info("node0 mined 100 blocks")
+        assert self.get_best_block(self.nodes[0])['height'] == 100
+
+        # endorse block 100 (fork A tip)
+        addr0 = self.nodes[0].getnewaddress()
+        txid = endorse_block(self.nodes[0], self.apm, 100, addr0)
+        self.log.info("node0 endorsed block 100 (fork A tip)")
+        # mine pop tx on node0
+        self.log.info('Should not accept POP data before activation block height')
+        assert_raises_rpc_error(-1, 'block contains PopData before PopSecurity has been enabled', lambda: self.nodes[0].generate(nblocks=1))
+
+        self.log.warning("_cannot_endorse() succeeded!")
+
+    def _can_endorse(self):
+        self.log.warning("starting _can_endorse()")
+        connect_nodes(self.nodes[0], 1)
+        self.sync_all()
+        disconnect_nodes(self.nodes[1], 0)
+
+        mine_until_pop_enabled(self.nodes[0])
+        lastblock = self.nodes[0].getblockcount()
+
+        # endorse block 200 (fork A tip)
+        addr0 = self.nodes[0].getnewaddress()
+        txid = endorse_block(self.nodes[0], self.apm, lastblock, addr0)
+        self.log.info("node0 endorsed block {} (fork A tip)".format(lastblock))
+        # mine pop tx on node0
+        self.nodes[0].generate(nblocks=1)
+        tip = self.get_best_block(self.nodes[0])
+        self.log.info("node0 tip is {}".format(tip['height']))
+
+        self.nodes[1].generate(nblocks=250)
+        tip2 = self.get_best_block(self.nodes[1])
+        self.log.info("node1 tip is {}".format(tip2['height']))
+
+        connect_nodes(self.nodes[0], 1)
+        self.sync_all()
+        bestblocks = [self.get_best_block(x) for x in self.nodes]
+        assert_equal(bestblocks[0]['hash'], bestblocks[1]['hash'])
+        self.log.info("all nodes switched to common block")
+
+        for i in range(len(bestblocks)):
+            assert bestblocks[i]['height'] == tip['height'], \
+                "node[{}] expected to select shorter chain ({}) with higher pop score\n" \
+                "but selected longer chain ({})".format(i, tip['height'], bestblocks[i]['height'])
+
+        self.log.info("all nodes selected fork A as best chain")
+        self.log.warning("_can_endorse() succeeded!")
+
+    def run_test(self):
+        """Main test logic"""
+        from pypopminer import MockMiner
+        self.apm = MockMiner()
+        self._cannot_endorse()
+        self.restart_node(0)
```

Comment:
why do you restart here?

Reply:
Because bad endorsement stays in the mempool. Mining a block would fail.